habitat_code/kobylarz_kmeans_review_code_20240313.py

```diff
@@ -0,0 +1,603 @@
+#      ```
+#   2. Modify the following sample code to search for granules of the HLSL30 product overlapping the watershed boundary from May to October 2023 (there should be 76 granules):
+#      ```python
+#      results = earthaccess.search_data(
+#          short_name="...",
+#          cloud_hosted=True,
+#          bounding_box=tuple(gdf.total_bounds),
+#          temporal=("...", "..."),
+#      )
+#      ```
+
+# In[5]:
+
+
+# Search multispectral data from Earthdata
+
+earthaccess.login(persist=True)
+
+results = earthaccess.search_data(
+    short_name="HLSL30",
+    cloud_hosted=True,
+    bounding_box=tuple(huc_gdf.total_bounds),
+    temporal=("2023-05-01", "2023-09-30"),
+)
+
+
+# ### Compile information about each granule
+# 
+# I recommend building a GeoDataFrame, as this will allow you to plot the granules you are downloading and make sure they line up with your shapefile. You could also use a DataFrame, dictionary, or a custom object to store this information.
+# 
+# **YOUR TASK -- Write a function that:**
+#   1. For each search result:
+#       1. Get the following information (HINT: look at the ['umm'] values for each search result):
+#           - granule id (UR)
+#           - datetime
+#           - geometry (HINT: check out the shapely.geometry.Polygon class to convert points to a Polygon)
+#       2. Open the granule files. I recomment opening one granule at a time, e.g. with (`earthaccess.open([result]`).
+#       3. For each file (band), get the following information:
+#           - file handler returned from `earthaccess.open()`
+#           - tile id
+#           - band number
+#   2. Compile all the information you collected into a GeoDataFrame
+
+# In[5]:
+
+
+# Open the Earthaccess results so you only have to do it once
+@cached("open_results", False)
+def open_results_and_cache(results):
+    open_results = earthaccess.open(results)
+    return open_results
+open_results = open_results_and_cache(results)
+
+
+# In[7]:
+
+
+# Process the granules from Earthdata
+def create_granule_df(earthdata_results, open_results):
+    """
+    Process the granules returned by an Earthdata search
+
+    Parameters
+    ==========
+    earthdata_results : list
+      A list object returned by earthacces.search_data().
+      This contains the result granules from the Earthdata search.
+      
+    open_results : opened earthdata_results object
+
+    Returns
+    =======
+    granule_df : DataFrame
+      A dataframe with a row for each granule file.
+      Each row will contain the granule ID, the download URL,
+      the datetime, the tile ID, and the band.
+    """
+    columns = ['geometry','granule_id', 'file_url', 'tile_id', 'datetime', 'band_number']
+    granule_df = gpd.GeoDataFrame(columns=columns)
+    # Loop through each file in the granule results and accumulate attributes
+    for url in open_results:
+        file_url = str(url.url)
+        # Should probably use regular expression here
```

Comment:
Yeah, that would be my first choice. You can also check out `.split()`, like @eggvoice did, to avoid having to reprocess the Fmask band name.